src/util/ConfigManager/ConfigManager.h

```diff
@@ -0,0 +1,247 @@
+    }
+  }
+
+  /*
+  @brief Creates and adds a new configuration option.
+
+  @tparam OptionType The type of value, the configuration option can hold.
+
+  @param optionName Name of the option. Additionally, this the one name is the
+  entire path in json, needed to address the value held by this option.
+  @param optionDescription A description for the configuration option.
+  @param variableToPutValueOfTheOptionIn The value held by the configuration
+  option will be copied into this variable, whenever the value in the
+  configuration option changes.
+  @param defaultValue A default value for the configuration option. If none is
+  given, signified by an empty optional, then a value for the configuration
+  option MUST be given at runtime.
+  */
+  template <typename OptionType>
+  requires ad_utility::isTypeContainedIn<OptionType,
+                                         ConfigOption::AvailableTypes>
+  void createConfigOption(std::string_view optionName,
+                          std::string_view optionDescription,
+                          OptionType* variableToPutValueOfTheOptionIn,
+                          std::optional<OptionType> defaultValue =
+                              std::optional<OptionType>(std::nullopt)) {
+    createConfigOption<OptionType>(
+        VectorOfKeysForJson{std::string{optionName}}, optionDescription,
+        variableToPutValueOfTheOptionIn, defaultValue);
+  }
+
+  /*
+  @brief Get all the added configuration options.
+  */
+  const std::vector<ConfigOption>& getConfigurationOptions() const;
+
+  /*
+  @brief Sets the configuration options based on the given json. Note: This will
+  overwrite values held by the configuration data, if there were values for them
+  given inside the string.
+
+  @param j There will be an exception thrown, if:
+  - `j` doesn't contain values for all configuration options, that must be set
+  at runtime.
+
+  - Same, if there are values for configuration options, that do not exist.
+
+  - `j` is anything but a json object literal.
+  */
+  void parseConfig(const nlohmann::json& j);
+
+  /*
+  @brief Parses the given short hand and returns it as a json object,
+   that contains all the described configuration data.
+
+  @param shortHandString The language of the short hand is defined in
+  `generated/ConfigShorthand.g4`.
+  */
+  static nlohmann::json parseShortHand(const std::string& shortHandString);
+
+  /*
+  @brief Returns a string containing a json configuration and the
+  string representations of all added configuration options.
+
+  @param printCurrentJsonConfiguration If true, the current values of the
+  configuration options will be used for printing the json configuration. If
+  false, an example json configuration will be printed, that uses the default
+  values of the configuration options, and, if an option doesn't have a default
+  value, a dummy value.
+  */
+  std::string printConfigurationDoc(bool printCurrentJsonConfiguration) const;
```

Comment:
I implemented this, but then I had a realization: Doesn't that destroy any usage of `printConfigurationDoc(true)` for error message creation?

I mean, currently printing the current values of all configuration options is mainly used for generating error messages, not much more. But in that situation, not all `ConfigOptions` are guaranteed to be set, which causes the the generation of the error messages to replace the wanted error message with the error of not all config options holding valid values.

And outside of that, for what can you really use this function? The user already has the information, to what the values were set, **because he sets them.** Printing them out again is kinda redundant...

Reply:
Thanks for pointing this out. It should still be safe (never undefined behavior) to call this with true, ESPECIALLY when we have partial parsing.
This can be done by (in the `true` case) reading the default values from their separate storage, and access the actual storage only if `wasSetAtRuntime` is true. 

And I somewhat diesagree with the **because he sets them**:
Often you start something and then you have a configuration that is partially from default values, partially from a config file (which you maybe have copied from somewhere) and partially changed via shortHand configuration or something else. Then having the complete actually used config prominently at the beginning of e.g. your Logfile is very very useful and I definitely want to use it.

In fact we currently have many manual "You set tthe parameter Bla to  value Blubb for this Index" log messages, that can be replaced by this function.

Reply:
> Thanks for pointing this out. It should still be safe (never undefined behavior) to call this with true, ESPECIALLY when we have partial parsing.
> This can be done by (in the `true` case) reading the default values from their separate storage, and access the actual storage only if `wasSetAtRuntime` is true.

A small reminder: If a `ConfigOption` has a default value, the variable, that the variable pointer points to, gets set to that default value in the constructor.
Additionally, what should we do, if a `ConfigOption` was never set **and** also doesn't have a default value? Would a `Value was never initialized`, instead of a value, suffice?

Reply:
TODO: This [issue](https://github.com/ad-freiburg/qlever/pull/977#discussion_r1235104298) is only doable, after hashing this point out.

Reply:
I think `"value was never initialized"` (small "v") in this case is fine. This should typically only occur when we have the errors during parsing. Does this hash out the point?:)

Reply:
Yes, it does. Thank you.